### ISSUE TITLE
Make FAIL_ON_NULL_FOR_PRIMITIVES apply to primitive arrays and other types that wrap primitives

### ISSUE THREAD

**Issue by harleensahni** [FasterXML/jackson-databind#403]
`DeserializationFeature.FAIL_ON_NULL_FOR_PRIMITIVES` only seems to work on primitive properties on beans. It doesn't seem to work on reading straight values (I don't care that much about this), and reading primitive arrays. 

In addition, it would be good if this functionality, or a new DeserializationFeature was added to support failing on nulls in Lists that store wrap primitive types since people often use List<Integer> as a property type instead of int[] for convenience. 

Here is a test case I wrote to demonstrate this:  

``` java
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.JsonMappingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.Test;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;

public class TestNullPrimitives {

    public static class PrimitiveTest{
        public int value;
        public int[] valueArray;
        public List<Integer> valueList;
    }

    @Test(expected = JsonMappingException.class)
    public void testReadNullInt() throws IOException {

        ObjectMapper mapper = new ObjectMapper();
        mapper.configure(DeserializationFeature.FAIL_ON_NULL_FOR_PRIMITIVES, true);
        // Reads value as 0
        int value  = mapper.readValue("null", int.class);
        System.out.println("value:" + value);

    }

    @Test(expected = JsonMappingException.class)
    public void testObjectReadNullProperty() throws IOException {
        ObjectMapper mapper = new ObjectMapper();
        mapper.configure(DeserializationFeature.FAIL_ON_NULL_FOR_PRIMITIVES, true);
        // throws exception
        PrimitiveTest testObject = mapper.readValue("{\"value\":null}", PrimitiveTest.class);
    }
    @Test(expected = JsonMappingException.class)
    public void testObjectReadNullArrayEntryInProperty() throws IOException {
        ObjectMapper mapper = new ObjectMapper();
        mapper.configure(DeserializationFeature.FAIL_ON_NULL_FOR_PRIMITIVES, true);
        // Reads valueArray as [0]
        PrimitiveTest testObject = mapper.readValue("{\"valueArray\":[null]}", PrimitiveTest.class);
        System.out.println("Contents:" + Arrays.toString(testObject.valueArray));
        System.out.println("Size:" + testObject.valueArray.length);
    }

    @Test(expected = JsonMappingException.class)
    public void testObjectReadNullListEntryInProperty() throws IOException {
        ObjectMapper mapper = new ObjectMapper();
        mapper.configure(DeserializationFeature.FAIL_ON_NULL_FOR_PRIMITIVES, true);
        // Reads valueList as [null]
        PrimitiveTest testObject = mapper.readValue("{\"valueList\":[null]}", PrimitiveTest.class);
        System.out.println("Contents:" + testObject.valueList);
        System.out.println("Size:" + testObject.valueList.size());
    }
}
```

**Comment by cowtowncoder** [comment]
One comment here; I do not think this feature should fail for `List<Integer>`, as `Integer` is not primitive but a wrapper. While I understand while it sometimes would be nice (given that Java does not allow `List<int>`), it would be misnomer, and specifically confusing if it didn't apply to stand-alone wrapper values. Further, if it did apply to stand-alone values, it would be unintuitive as major usage reason for `Integer` over `int` is specifically to allow nulls.

Anyway; I consider this to only apply for `int[]` etc.

**Comment by shmosel** [comment]
Any update on the primitive array bug? Or can anyone suggest a workaround to force it to fail?

**Comment by cowtowncoder** [comment]
@shmosel I try to update issues with notes when there is progress so no updates. But as luck would have it I did work in related area for `2.9.0.pr1` so I hope to have a look at this relatively soon (for 2.9.0.pr2 at least -- not sure how safe behavioral change this would be for 2.8)

**Comment by cowtowncoder** [comment]
Implemented for 2.9.0(.pr2), added tests: should cover both coercion from literal `null` and empty String, similar to how it works for primitive-valued POJO properties.
